### ISSUE TITLE
CBA_fnc_notify error sometimes pops up during combat

### ISSUE THREAD

**Issue by GoldJohnKing** [Vdauphin/HeartsAndMinds#936]
**Arma 3 Version:** `1.98` (stable)
**CBA Version:** `3.15.1` (stable)
**ACE3 Version:** `3.13.3` (stable)
**RHS USAF Version:** `0.5.2` (stable)
**Hearts and minds Version:** `master_daily` (c8d0e82)

**Mods:**
```
- CBA_A3
- ace
- RHS USAF
- RHS AFRF
- RHS GREF
- RHS SAF
- CUP Terrains - Core
- Kujari
- USAF MOD Main Pack
- USAF MOD Utility pack
- USAF MOD Fighters Pack
- JSRS SOUNDMOD
```

**Description:**
- This error sometimes pops up during combat, I think this might related to `btc_fnc_show_hint`.
```
15:48:50 Error in expression <nc_colorRGBAtoHTML;

private _isImage = toLower _text select [count _text - 4] i>
15:48:50   Error position: <toLower _text select [count _text - 4] i>
15:48:50   Error tolower: 类型阵列，期望类型字串
15:48:50 File \x\cba\addons\ui\fnc_notify.sqf [CBA_fnc_notify]..., line 5879
```
- The Chinese characters above should be `Type Array, String Expected` in English.

**Steps to reproduce:**
- This sometimes pops up during combat.

**Comment by Vdauphin** [comment]
Please trow the .rtp .pbo ...

**Comment by GoldJohnKing** [comment]
https://github.com/Vdauphin/HeartsAndMinds/files/5040174/BTC.co%4030_Hearts_and_Minds.tem_kujari.zip

https://github.com/Vdauphin/HeartsAndMinds/files/5040194/Arma3_x64_2020-08-07_14-03-03.zip

**Comment by Vdauphin** [comment]
Do you know which hint is shown or should be shown when the error occur ?

**Comment by GoldJohnKing** [comment]
> Do you know which hint is shown or should be shown when the error occur ?

Unfortunately, no. There are so many players in the server...

I bet case 20 or 21 in `btc_fnc_show_hint`, since this never happens before "show reputation change" feature was introduced, and the error seems to be the same when `_content` or `_lineN` was passed incorrectly or `null` in `CBA_fnc_notify`.

I noticed `_player` sometimes pass `Error: no unit`, may be related?

**Comment by Vdauphin** [comment]
log all parameters passed to `btc_fnc_show_hint` then we will know what create the issue

> I noticed _player sometimes pass Error: no unit, may be related? If killer is not player, such as AI and collision, will _player pass correctly?

Don't think

**Comment by GoldJohnKing** [comment]
> log all parameters passed to btc_fnc_show_hint

May I ask how to...? `btc_debug` seems doing nothing this...

**Comment by Vdauphin** [comment]
Just add this:
```sqf
[format ["%1", _this], __FILE__, [false]] call btc_fnc_debug_message;
```
here: https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/common/show_hint.sqf#L25

**Comment by GoldJohnKing** [comment]
Finally! I've captured the issue in rpt. At the exact time, all players wondering around at base, no fire, no respawn, no units killed or vehicles respawn.

```
16:32:24 (3:59:00) fnc\common\show_hint - [20,[[1,1,1],any]]
16:32:24 Error in expression <nc_colorRGBAtoHTML;

private _isImage = toLower _text select [count _text - 4] i>
16:32:24   Error position: <toLower _text select [count _text - 4] i>
16:32:24   Error tolower: 类型阵列，期望类型字串
16:32:24 File \x\cba\addons\ui\fnc_notify.sqf [CBA_fnc_notify]..., line 5879
```
(The Chinese characters above should be Type Array, String Expected in English.)

In case you need the full rpt file:
[Arma3_x64_2020-08-14_16-18-47.zip](https://github.com/Vdauphin/HeartsAndMinds/files/5073475/Arma3_x64_2020-08-14_16-18-47.zip)

I think it should be related to btc_fnc_rep_notify.

And please notice I've edited the show_hint and added case 22 and case 23 to show which player killed friendly vehicles, and I've commented out this:

https://github.com/Vdauphin/HeartsAndMinds/blob/c6a29ff15b98a67ec65cc6e6cfe413258972533b/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/eh/player.sqf#L34
and set this to zero:

https://github.com/Vdauphin/HeartsAndMinds/blob/c6a29ff15b98a67ec65cc6e6cfe413258972533b/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/def/mission.sqf#L676

I'm pretty sure they have nothing to do with the issue.

**Comment by Vdauphin** [comment]
> I think it should be related to btc_fnc_rep_notify.

I don't understand why it return `any`, should be `objNull` at least (https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/rep/change.sqf#L26)
Add:
```sqf
[format ["GLOBAL %1 - CHANGE %2 - _player %3", btc_global_reputation, _reputation, _player], __FILE__, [btc_debug, btc_debug_log]] call btc_fnc_debug_message;
``` 
here
https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/rep/change.sqf#L30
The `_reputation` number should give us the event passing `any` as `_player` so it will be faster to reproduce the error

Don't forget to enable debug log and trow rpt server and player side

_____________________________________
Also your RPT show 
```sqf
16:22:24 (3:48:55) fnc\common\show_hint - [21,[[1,0,0],200beebe040# 1806240: ah64d.p3d REMOTE]]
16:22:24 WARNING: Function 'name' - 200beebe040# 1806240: ah64d.p3d REMOTE has no unit
``` 
Looks like an Apache kill something or crash?  Do you remember? Is it due to your changes?
For an unknown reason the vehicle is passed as instigator which should not be the case: instigator: Object - Person who pulled the trigger (https://community.bistudio.com/wiki/Arma_3:_Event_Handlers#Killed)

**Comment by GoldJohnKing** [comment]
> Don't forget to enable debug log and trow rpt server and player side

Okay, I added it and I'll upload the contents once I captured it.

> Looks like an Apache kill something or crash? Do you remember? Is it due to your changes?

At that time a player crashed an Apache into a tree... But no, this should not caused by my modification. I just added a function to show which player killed or crashed btc_veh & btc_helo as case 22 & 23 to `show_hint.sqf`. It's true that vehicle sometimes is passed as instigator, however I use an if to detect if instigator is a player, so it functions well as it will exit if instigator is a vehicle or enemy.

You can see here:

https://github.com/GoldJohnKing/HeartsAndMinds/blob/cb3df5dcc84bf2d992b9c3f027f5ce1a6f27efdc/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/eh/veh_add_respawn.sqf#L44-#L56

and here:

https://github.com/GoldJohnKing/HeartsAndMinds/blob/cb3df5dcc84bf2d992b9c3f027f5ce1a6f27efdc/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/common/show_hint.sqf#L118-#L137

**Comment by GoldJohnKing** [comment]
Ah, forget to mention...

The error occurred after the Apache crashed, then the player drives a UH-60M to sling load it, and then crash it again near base, then he respawn, and drive on another helicopter, then the error occurred.

Because there're no other players outside base at that time, so I think the error should not related to player activity. I prefer if it's caused by a civilian vehicle crashed into another civilian, or something like that.

**Comment by GoldJohnKing** [comment]
I did not found something strange at server rpt at the same time client pops up the error. I'm not sure if I lost something.

[rpt, setver and client.zip](https://github.com/Vdauphin/HeartsAndMinds/files/5080625/rpt.setver.and.client.zip)

**Comment by Vdauphin** [comment]
Not really sure, I need to test some stuff may be due to something like this:
```sqf
[-1, objNull] call btc_fnc_rep_change;
[1, objNull] call btc_fnc_rep_change;
```
And if you wait 10min the bug should appear.

Or is due to `btc_rep_malus_player_respawn = 0; ` but it shouldn't because there is no log about that in your RPT

I will take a look when I am back home

**Comment by GoldJohnKing** [comment]
> Or is due to btc_rep_malus_player_respawn = 0; but it shouldn't because there is no log about that in your RPT

You see I've also commented out here: 
https://github.com/Vdauphin/HeartsAndMinds/blob/c6a29ff15b98a67ec65cc6e6cfe413258972533b/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/eh/player.sqf#L34
So I don't think this is the issue.

> And if you wait 10min the bug should appear.

I'll do that test at my side, too. But I don't get it... Why I need to wait 10min?

**Comment by Vdauphin** [comment]
> You see I've also commented out here:

If you didn't commented, you would have the bug more frequently I think

> I don't get it... Why I need to wait 10min?

When the reputation change is between -3 and 3, the H&M wait until is above by checking `btc_rep_delayed`  (https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/rep/change.sqf#L39)
When you did enough stuff to change the reputation outside -3 and 3, the reputation notification is triggered.
But there is also a time out of 10min (https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/rep/change.sqf#L42)
So if the reputation doesn't change outside -3 and 3, the reputation change is show after 10min.
At the end of both, `btc_rep_delayed` is reset (https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/rep/notify.sqf#L47)
The objective is to not show reputation every time you kill an enemy for example.

I think the "loop" (https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/rep/change.sqf#L38-L44) is started twice at the same time for a unknown reason.
So when the first loop has done its job, the second is waiting but after 10min end and show the reputation stored in `btc_rep_delayed` which is `[0, []]` due to the previous reset.
That with we see ` [[1,1,1],any]`, `[1,1,1]` mean a reputation of `0` and `any` mean the array with player is empty.

**Comment by GoldJohnKing** [comment]
Ah, I totally get it now. I didn't noticed that is a loop...

I will try and see if this solves the issue.

**Comment by GoldJohnKing** [comment]
Issue seems to be fixed. The error message has not popped up so far.

**Comment by GoldJohnKing** [comment]
I believe this is solved by #942.
